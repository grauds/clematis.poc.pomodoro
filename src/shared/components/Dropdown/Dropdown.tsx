import React, { useEffect, useRef } from 'react';
import { noop } from '../../../utils/noop';

import styles from './dropdown.css';

interface IDropdownProps {
  button: React.ReactNode;
  children: React.ReactNode;
  isOpen?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}

export function Dropdown({
  button,
  children,
  isOpen,
  onOpen = noop,
  onClose = noop,
}: Readonly<IDropdownProps>): React.JSX.Element {
  const ref = useRef<HTMLDivElement>(null);
  const [isDropdownOpen, setIsDropdownOpen] = React.useState(isOpen);
  useEffect(() => setIsDropdownOpen(isOpen ?? false), [isOpen]);
  useEffect(() => (isDropdownOpen ? onOpen() : onClose()), [isDropdownOpen]);

  useEffect(() => {
    function handleClick(event: MouseEvent) {
      if (
        event.target instanceof Node &&
        !ref.current?.contains(event.target)
      ) {
        onClose?.();
        setIsDropdownOpen(false);
      }
    }
    document.addEventListener('click', handleClick);

    return () => {
      document.removeEventListener('click', handleClick);
    };
  }, []);

  const handleOpen = () => {
    if (isOpen === undefined) {
      setIsDropdownOpen(!isDropdownOpen);
    }
  };

  return (
    <div className={styles.container} ref={ref}>
      <button aria-label='menuButton' onClick={handleOpen}>{button}</button>
      {isDropdownOpen && (
        <div className={styles.listContainer}>
          <button className={styles.list} onClick={() => setIsDropdownOpen(false)}>
            {children}
          </button>
        </div>
      )}
    </div>
  );
}
