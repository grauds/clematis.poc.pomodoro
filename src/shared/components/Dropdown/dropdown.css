.container {
  width: 100%;
  flex: 1 1 0%;
}

.listContainer {
  position: relative;
  z-index: 1000;
}

.list {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
}
