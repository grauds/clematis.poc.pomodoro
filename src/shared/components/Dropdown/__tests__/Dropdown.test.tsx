import React from "react";
import { Dropdown } from "../Dropdown";
import { shallow } from "enzyme";

describe("Dropdown", () => {
  test("should render", () => {
    const wrapper = shallow(
      <Dropdown button={<button />}>
        <div />
      </Dropdown>
    );
    expect(wrapper).toBeDefined();
    console.log(wrapper.find("div.container").debug());
    expect(wrapper.find("div.container").isEmptyRender()).toBeFalsy();
  });

  test("should render (snapshot)", () => {
    const wrapper = shallow(
      <Dropdown button={<button />}>
        <div />
      </Dropdown>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
