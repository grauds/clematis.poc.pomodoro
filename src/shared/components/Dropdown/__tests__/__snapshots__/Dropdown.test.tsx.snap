// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
<div
  class="container"
>
  <button
    aria-label="menuButton"
  >
    <button />
  </button>
</div>
`;
